### PR TITLE
fix: handle beforeLoad throwing notFound correctly

### LINKED ISSUE #6779: Hydration error when throwing notFound() in beforeLoad of a child route
### Which project does this relate to?

Start

### Describe the bug

Throwing `notFound()` from beforeLoad within a child route leads to a hydration error.

```
Uncaught Error: Hydration failed because the server rendered HTML didn't match the client. As a result this tree will be regenerated on the client. This can happen if a SSR-ed Client Component used:

- A server/client branch `if (typeof window !== 'undefined')`.
- Variable input such as `Date.now()` or `Math.random()` which changes each time it's called.
- Date formatting in a user's locale which doesn't match the server.
- External changing data without sending a snapshot of it along with the HTML.
- Invalid HTML tag nesting.

It can also happen if the client has a browser extension installed which messes with the HTML before React loaded.

https://react.dev/link/hydration-mismatch

  ...
    <RouterProvider router={{...}}>
      <RouterContextProvider router={{...}}>
        <Matches>
          <SafeFragment fallback={null}>
            <Transitioner>
            <MatchesInner>
              <CatchBoundary getResetKey={function getResetKey} errorComponent={function ErrorComponent} ...>
                <CatchBoundaryImpl getResetKey={function getResetKey} onCatch={function}>
                  <MatchImpl matchId="__root__/">
                    <RootDocument>
                      <html>
                        <head>
                          <HeadContent>
                            <Asset>
                            <Asset>
                            <Asset>
                            <Asset>
                            <Asset>
                            <Asset>
                            <Asset>
                            <Asset>
                            <Asset>
                            <Asset>
                            <Asset>
                            <Asset>
                            <Asset tag="link" attrs={{rel:"style...", ...}} nonce={undefined}>
+                             <link
+                               rel="stylesheet"
+                               href="/src/styles/app.css"
+                               nonce={undefined}
+                               suppressHydrationWarning={true}
+                             >
-                             <script src="https://codesandbox.io/p/chrome-devtool/protocol/index.js">
                            ...
                        ...

    at throwOnHydrationMismatch (react-dom_client.js?v=07a9c313:3918:13)
    at beginWork (react-dom_client.js?v=07a9c313:8692:196)
    at runWithFiberInDEV (react-dom_client.js?v=07a9c313:995:72)
    at performUnitOfWork (react-dom_client.js?v=07a9c313:12559:98)
    at workLoopConcurrentByScheduler (react-dom_client.js?v=07a9c313:12555:11)
    at renderRootConcurrent (react-dom_client.js?v=07a9c313:12537:71)
    at performWorkOnRoot (react-dom_client.js?v=07a9c313:11764:152)
    at performWorkOnRootViaSchedulerTask (react-dom_client.js?v=07a9c313:13503:9)
    at MessagePort.performWorkUntilDeadline (react-dom_client.js?v=07a9c313:34:50)
```

### Your Example Website or App

https://codesandbox.io/p/devbox/friendly-haze-v4nyyq

### Steps to Reproduce the Bug or Issue

1. Visit the a posts page (https://v4nyyq-3000.csb.app/posts/abc)
2. See an initial lack of styles applied and check the console

### Expected behavior

I would expect to see the styles applied on first paint 

### Screenshots or Videos

_No response_

### Platform

- Router / Start Version: 1.163.2

### Additional context

in a child route - 
throwing error from loader ✅ 
throwing notFound from loader ✅ 
throwing error from beforeLoad ✅ 
throwing notFound from beforeLoad ❌

### PR DESCRIPTION
**Pull request by schiller-manuel**
fixes #6779

## Summary by CodeRabbit

* **New Features**
  * Enhanced not-found behavior: beforeLoad can trigger targeted notFound boundaries (including root) while preserving parent loader data; per-route notFoundComponent support; new demo routes and navigation to exercise deep and parent-boundary cases; exported rootRouteId for targeted notFound.

* **Documentation**
  * Updated guide and migration example to explain boundary resolution and show notFoundComponent usage instead of prior API.

* **Tests**
  * Expanded E2E and unit tests across frameworks for beforeLoad not-found, routeId targeting, deep/parent-boundary scenarios, and hydration global-not-found checks.

### CODE REVIEW

File: packages/router-core/src/load-matches.ts [hunk 1]
```diff
@@ -906,9 +890,8 @@ export async function loadMatches(arg: {
   updateMatch: UpdateMatchFn
   sync?: boolean
 }): Promise<Array<MakeRouteMatch>> {
-  const inner: InnerLoadContext = Object.assign(arg, {
-    matchPromises: [],
-  })
+  const inner: InnerLoadContext = Object.assign(arg)
```
Comment (Sheraff) [inline packages/router-core/src/load-matches.ts]: This line doesn't do anything anymore, it's just 
```suggestion
  const inner: InnerLoadContext = arg
```

File: packages/router-core/src/load-matches.ts [hunk 2]
```diff
@@ -919,77 +902,159 @@ export async function loadMatches(arg: {
...
+  } catch {
+    const settled = await Promise.allSettled(matchPromises)
+
+    for (const result of settled) {
+      if (result.status !== 'rejected') continue
+
+      const reason = result.reason
+      if (!firstRedirect && isRedirect(reason)) {
+        firstRedirect = reason
+      }
```
Comment (Sheraff) [inline packages/router-core/src/load-matches.ts]: We could break right here, right? if we have a redirect and no notFound (yet) no need to continue?
```suggestion
      if (!firstRedirect && isRedirect(reason)) {
        throw reason
      }
```

File: packages/router-core/src/load-matches.ts [hunk 3]
```diff
@@ -919,77 +902,159 @@ export async function loadMatches(arg: {
...
 
-    const readyPromise = triggerOnReady(inner)
-    if (isPromise(readyPromise)) await readyPromise
-  } catch (err) {
-    if (isNotFound(err) && !inner.preload) {
-      const readyPromise = triggerOnReady(inner)
-      if (isPromise(readyPromise)) await readyPromise
-      throw err
+  if (firstRedirect) {
+    throw firstRedirect
```
Comment (Sheraff) [inline packages/router-core/src/load-matches.ts]: Should the redirection be honored if it was found "under" a `notFound`? I feel like whichever first error we see (notFound or redirect) is the one that should be used

Comment (tannerlinsley) [reply packages/router-core/src/load-matches.ts]: I think first one in wins, especially here.

File: packages/router-core/src/load-matches.ts (hunk 2 above)
Comment (tannerlinsley) [reply packages/router-core/src/load-matches.ts]: I think so yes.

File: packages/router-core/src/load-matches.ts [hunk 4]
```diff
@@ -74,64 +73,45 @@ const buildMatchContext = (
...
-    // Ensure we have a notFoundComponent
-    invariant(
-      routeCursor.options.notFoundComponent,
-      'No notFoundComponent found. Please set a notFoundComponent on your route or provide a defaultNotFoundComponent to the router.',
-    )
-  }
-
-  // Find the match for this route
-  const matchForRoute = inner.matches.find((m) => m.routeId === routeCursor.id)
+  const requestedRouteId = err.routeId ? String(err.routeId) : undefined
```
Comment (Sheraff) [inline packages/router-core/src/load-matches.ts]: Why `String(err.routeId)` here? what else could it be if not already a string?

File: packages/router-core/src/load-matches.ts (hunk 3 above)
Comment (schiller-manuel) [reply packages/router-core/src/load-matches.ts]: this is the scenario where a loader throws a `redirect` after a child beforeLoad threw a `notFound`

i think we cannot honor the `notFound` since there is no loader data for this route and so we cannot render its route component (the types would not make the loader data optional if you throw a `redirect`)
